themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md

```diff
@@ -0,0 +1,87 @@
+* IaC Recommended Practices: Refactoring for Reuse
+
+## Reviewing Zephyr's current status
+
+As this post is somewhat of a continuation of the previous post, Zephyr's use of Pulumi remains the same as it did at the end of the third post. That is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect or portion of the overall set of requirements for Zephyr's online store:
+
+* The "zephyr-infra" project handles base infrastructure (VPC and related constructs).
+* The "zephyr-k8s" project handles the Kubernetes platform layer, building atop the infrastructure components provided by the "zephyr-infra" project.
+* The "zephyr-app" project handles _only_ the deployment of the online store services onto Kubernetes, layering on top of the platform provided by the "zephyr-k8s" project.
+
+{{% notes %}}
+You can continue to use the `multi-project` tag in the GitHub repositories to see the state of Zephyr's code and projects for this blog post.
+{{% /notes %}}
+
+It's clear that there are key dependencies across projects. The "zephyr-k8s" project needs to know the VPC ID and the subnet IDs from the base infrastructure project. Similarly, the "zephyr-app" project needs to have access to the Kubernetes cluster details from the "zephyr-k8s" project. As the Zephyr team went about building the code that is now in use in their multi-project architecture, they needed to determine how best to handle these cross-project dependencies. The Zephyr team knew that hard-coding output values from one project as configuration values in another stack was not the ideal way (Pulumi **strongly** recommends against this approach), but what was the best solution to use? The answer to that question is _stack references_.
+
+## Linking stacks
+
+For those that are unfamiliar, stack references are a feature in Pulumi that allows you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [here in the Pulumi docs](/docs/intro/concepts/stack/#stackreferences).
+
+Using stack references, the Zephyr team could make the necessary connections they needed:
+
+* The code handling the Kubernetes cluster could reference the VPC ID and subnet IDs from the base infrastructure layer.
+* Similarly, the code that was responsible for deploying the online store services could reference the necessary Kubernetes connection details for the Kubernetes cluster.
+
+It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Fortunately, adding stack outputs after the fact is easy and generally has no impact whatsoever on existing infrastructure.
```

Comment:
We should either omit "generally" or explain the exceptions here. (I believe it never has an effect, but maybe there's something I'm not remembering.)

Reply:
Done, taken care of in f28ec13c